projecttacoma/measure-repository
# Add clone operation for draft artifacts

# Summary
This PR implements the `$clone` operation as described by the Quality Measure IG [here](https://build.fhir.org/ig/HL7/cqf-measures/measure-repository-service.html#authoring-measure-repository). There is currently no OperationDefinition for $clone like there is for $draft and $release in the [CRMI IG](https://build.fhir.org/ig/HL7/crmi-ig/artifact-repository-service.html), so the functionality for this operation is subject to change. For now, this PR implements functionality to clone a draft artifact in the draft-repository (authoring measure repository). 

## New behavior
There is now a "Clone" button amongst the "Review", "Edit", and "Delete" buttons for draft artifacts in the draft-repository frontend. When the user clicks on the "Clone" button, a confirmation modal appears notifying the user that the action will clone that artifact and any of its children. The act of cloning (for now) means that a draft artifact and its children are given a new ID and their versions are incremented (tried 10 times and errors out if it does not have a unique URL and version combination), but the URL and status: "draft" remains the same. Again, this is subject to change as the definition of clone adapts and is potentially discussed at the Connectathon. 

## Code changes
- `app/src/components/DeletionConfirmationModal.tsx` -> `app/src/components/ConfirmationModal.tsx` - made the `DeletionConfirmationModal` reusable for clone confirmation as well.
- `app/src/components/ResourceInfoCard.tsx` - made success and error notifications reusable for delete and clone, added trpc `cloneMutation`, `ConfirmationModal` for clone, and "clone" button.
- `app/src/pages/authoring/index.tsx` - `trpc.useContext()` is now deprecated so it has to be replaced with `useUtils()`.
- `app/src/server/db/dbOperations.ts` - add `batchCloneDraft` function.
- `app/src/server/trpc/routers/draft.ts` - consolidate code into helper function `getParentDraftArtifactAndChildren` and add `cloneParent` procedure.
- `app/src/server/trpc/routers/service.ts` - reusable `modifyResource` for draft and clone.
- `app/src/util/draftHelper.ts` - move reused code for getting a parent artifact and its children from `routers/draft.ts`.
- `app/src/util/modifyResourceFields.ts` - make `modifyResourceToDraft` reusable for both modifying to draft and modifying for clone.

# Testing guidance
- `npm run check:all`
- `cd service`
- `npm run db:loadBundle <path to bundle>` (recommended to use the one I edited for recursive delete testing.
- `cd ..`
- `npm run start:all`
- Try drafting a parent artifact and then try combinations of cloning them and deleting them and make sure that there is no unexpected behavior. Make sure the user cannot directly clone or delete a child artifact. 
- Make sure the formatting for the ResourceCards and their buttons were not messed up with the addition of the "clone" button.
- 
# Question for Reviewer
- I was randomly having a lot of trouble with getting the clone confirmation modal to close even after the clone mutation was successful. This was strange because I had no problem with the delete confirmation modal. I spent a ton of time on this and just decided to add the `setIsCloneConfirmationModalOpen(false)` and `setIsDeleteConfirmationModalOpen(false)`, but if you comment out both of those lines, you will see that the behavior is unchanged for delete but does not work for clone. I am very curious why this behavior is happening if anyone has any ideas!
